### ISSUE TITLE
Cluster size scalability question

### ISSUE THREAD

**Issue by melaraj2** [nats-io/nats-server#3533]
I see references to "a recommended cluster size" in the documentation.   This statement is incongruent with true horizontal scalability.  If you can't make a cluster of whatever size, then there is really no horizontal scalability.  I understand the need to keep jetstream streams to 3 replicas, but does that really have anything to do with the overall size of the cluster?  can we make a 10-server cluster, but make a policy to create streams that only live in 3 cluster members?

Along the same lines of scalability, are there any plans for jetstream stream partitioning?

**Comment by derekcollison** [comment]
Yes clusters themselves can be any size, although if they get too big we might recommend a super cluster (many clusters). This is how we setup NGS ourselves.

In terms of JetStream partitioning, we have the ability to map subjects into partitioned subjects transparently today, so you can utilize multiple streams that are partitioned.

We have also been considering segmented streams.

And lastly, if you are using JetStream as a KV, when you do a get from a KV any mirrors will also transparently share the load as of 2.9.

**Comment by melaraj2** [comment]
Excellent, thank you.  I think the documentation states that only 3 or 5 jetstreams-enabled servers should be in a cluster.  Perhaps, it should say that when creating a stream, it should have 3 or 5 replicas.  I have a cluster with 6 nat-servers, they are all identical, and all have jetstream enabled.  I want to make sure this is ok for production.

Also, would Synadia offer a managed clusters product in the future?  Similar to confluent and Kafka?

**Comment by derekcollison** [comment]
Yes that might be a better way to describe it. Most CPs only have 3 AZs which define availability, so having R5 makes no sense in that situation really.

Synadia offers NGS and of course you can extend NGS with your own servers, which can run JetStream. We are also starting to support managed clusters in customers CP resources.

**Comment by kruegernet** [comment]
@derekcollison What is the status of thought on segmented streams today?

**Comment by derekcollison** [comment]
Still on the fuzzy roadmap, with 2.11 there will be a new filestore impl that will allow us to get to PB. Right now we are good IMO generally for large streams in TBs with reasonable subject cardinality and interior deletes. I am currently working on the subject cardinality problem.

Once those sort themselves out I think we will revisit in earnest the concept of segmented streams.

**Comment by kruegernet** [comment]
Thanks for the update. I ask next because I believe it's related for a lot of users to segmenting: is there a plan to allow changes to partition counts of subject partitioning via the or API from client code rather than involving server reloads of conf?

**Comment by kruegernet** [comment]
Thanks for the update. I ask next because I believe it's related for a lot of users to segmenting: is there a plan to allow changes to partition counts of subject partitioning via the or API from client code rather than involving server reloads of conf?

**Comment by derekcollison** [comment]
Not following. JetStream is a system concept, not an individual server one per se. Yes the server operator gives some broad parameters on where to store, and how much etc.

**Comment by kruegernet** [comment]
Sorry, I believe I was misunderstanding the docs. Clearly if I can do:
`nats server mapping "foo.*.*" "foo.{{wildcard(1)}}.{{wildcard(2)}}.{{partition(3,1,2)}}"` from CLI, then I don't need to update server conf files and cause all the cluster nodes to reload to affect changes here to the partition count.

**Comment by derekcollison** [comment]
Ah apologies you are talking about subject mappings and partitioning. These can be in account JWTs to in decentralized config mode. So you just update the account JWT and push to the system.

**Comment by ripienaar** [comment]
And also per stream config since 2.10

**Comment by jnmoyne** [comment]
To resume what has been said and explain a bit more:

As of 2.10 there are now two places where you can do "partitioning" in NATS: you can do subject transformation as part of streams which allows you to do partitioning for the consumers of the stream (which is the functionality most people are looking for). And you can change the number of partitions at any time by updating the stream's config. In the more rare cases where you want to partition a message flow into partitioned streams (plural) then you need to do Core NATS subject transformation which if you do not want it to be part of the server config you can do with the operator JWT-based security model where the subject transforms are part of the account JWT and as Derek mentioned you can update the mapping at any time by generating a new JWT and pushing it to the resolver.

**Comment by bjorndm** [comment]
How many KV buckets can one node reasonably support? And one NATS network?  For my use case, which is storing private data, I'm testing and would like to create 1 million buckets. My idea is to give each user their own private bucket. 

But on MacOS but around 20000 KV buckets I'm getting some performance problems on the creation of the buckets. Also, the startup is very slow with 20000 buckets, because the buckets are restored in series. Arguably they should be restored concurrently, 100 buckets at the time or so.

**Comment by derekcollison** [comment]
Keys can contain multiple tokens. So you can put many users in a single KV with proper permissions (we know we need to make this easier, but it is possible today).

`uuid.>`

**Comment by bjorndm** [comment]
Yes , that could also work, but I have some problems with the permissions for such an approach. So I filed #5204

**Comment by williamstein** [comment]
> How many KV buckets can one node reasonably support? And one NATS network? For my use case, which is storing private data, I'm testing and would like to create 1 million buckets. My idea is to give each user their own private bucket.

I've been using NATS version 2.11.0 in production for **one week*, and the number of KV's and streams (basically, one per user) is already starting to be a problem in one surprise way -- server restart time.  After about a week of use I have roughly 33K streams:
```

                        Storage: 24 GiB
                         Memory: 0 B
                        Streams: 33,931
                      Consumers: 35,443
```
I currently have exactly one big nats-server running on fast hardware (48 cpu's, 196GB RAM, SSD PD).   A big surprise, which is going to require me re-architecting my system in production, is that it takes a *LONG TIME* to start the server. With just the above data, it takes over 10 minutes, and it seems like the time scales linearly with the number of streams.   (I tested snapshotting the disk then starting a clone, stopping it cleanly, starting it again, etc.)  Basically I see a lot of lines like this at startup:
```
[7657] 2025/04/20 15:47:45.075110 [INF]   Restored 17 messages for stream 'COCALC > KV_project-6482ea22-4043-4d98-9e35-3785990bf8aa' in 12ms 
[7657] 2025/04/20 15:47:45.075227 [INF]   Starting restore for stream 'COCALC > KV_project-64837ab2-513e-4657-929b-f75e91e8e27b'
```
and those 12ms's all add up, and it looks like these are done linearly -- it's using 1 cpu, not 48.  

Before, we were storing our data in postgreSQL, where we had millions of these types of records, but startup of the database server was fast, so I was a taken by surprise. A big difference that you must account for in architecting a system around NATS (as opposed to some other data stores) if you're doing this -- "My idea is to give each user their own private bucket." is that even with a clean shutdown and startup, the cost of startup appears to currently be O(number of buckets) with no way to improve this via parallelism (except probably with multiple nats servers).     When I have a million streams (which would happen soon enough), then it might easily take an entire day to do a clean startup of the server.  Even if I sharded all streams across 100 servers, the startup time for any of the servers is still going to just keep growing. 

For my use case, I'm going to implement a tiering strategy, so if a stream isn't used for a while, then it is exported from NATS to a file in cloud storage, then deleted entirely from NATS.  When that stream is about to be used, there will be a service that reads it from cloud storage back into NATS.   Without implementing something like this, it seems to me that NATS will be totally unusable for an application like you're describing, since presumably a 1-day startup time for nats-server is not acceptable.     There is longterm discussion about adding tiered storage to NATS itself [here](https://github.com/nats-io/nats-server/discussions/3772), but I can't wait; anyways, it sounds from that thread like what I'm adding is a pretty standard thing many users have to do: "

**CAVEAT:** I'm not a NATS developer or even paying customer (yet!). I'm just a random user and **superfan** of NATS. :-).  Maybe everything I'm saying above can be trivially addressed by adding some configuration option somewhere.  

PS: I'm also developing a new storage system using ZFS and hit **exactly the same problem** -- there is a per-zfs filesystem startup cost when importing a zpool, which means I also have to implementing a very similar archiving and tiered storage solution.  A difference with ZFS is that they just landed parallel import, so at least using a server with many cpu's speeds up import by a factor of the number of cpus, which doesn't seem to be the case with NATS (yet?).    In short, with both ZFS and NATS, there is a substantial "per unit" cost at startup time, even if the server is cleanly shutdown and the objects are not being actively used.  This feels very different than PostgreSQL, though perhaps there the units are just different. This is something that is critical to be aware of when building a solution around NATS, and I wish it would be highlighted in the docs/FAQ.

**Comment by jnmoyne** [comment]
Having one KV (or stream) per customer is ultimately not a good design pattern because sooner or later the overhead of having all those streams becomes a limitation. With NATS it's better to use less streams and rely more on the fact that keys in NATS KV are actually subjects, and have a hierarchy of `.` delimited tokens. So rather than give each user its own bucket you carve a subset of the keys (e.g. prepend `<user id>.` to the key), and then if each user has its own user credentials you can limit what keys a particular client can do gets on by only allowing publish of requests on the right $JS.API subjects. It's more scalable to have a few streams or buckets with millions of keys in them leveraging subject hierarchy than to have millions (one per user) of streams or buckets with just a few records per user.
